### ISSUE TITLE
Test ROL_NonlinearProblemTest_MPI_4 timing out in ATDM waterman CUDA 9.2 builds starting 2019-11-14

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#6329]
CC: @trilinos/rol , @rppawlo  (Trilinos Nonlinear Solvers Product Lead), @bartlettroscoe, @fryeguy52

<Checklist>
<???: Add label "client: ATDM">
<???: Add label "ATDM Sev: Blocker" (by default but could be other "ATDM Sev: XXX")>
<???: Add label "type: bug"?>
<???: Add label for affected packages (e.g. "pkg: MueLu", "pkg: Tpetra", "pkg: Kokkos", etc.)>
<???: Add label "PA: ???Project Area???" (e.g. "PA: Linear Solvers", "PA: Data Services")>
<???: Add milestone "Initial cleanup of new ATDM ..." or "Keep promoted ATDM ...">
<???: Once GitHub Issue is created, add entries for tests to TrilinosATDMStatus/*.csv files>

## Next Action Status

The test fails when run by itself in 'waterman' builds, adding the thes property `RUN_SERIAL` will not help.  This test is disabled in all ATDM Trilinos 'waterman' CUDA builds in PR #7407 and has been merged to 'atdm-nightly' in commit 4804b08.  Next: Waiting for confirmation on CDash that this test is disabled in ATDM Trilinos CUDA builds starting testing day 2020-05-21 ...

Putting in review.
<status-and-or-first-action>

## Description

As shown in [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-10-22&end=2019-11-20&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=status&compare3=61&value3=failed) and [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-10-22&end=2019-11-20&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman) the test:

* `ROL_NonlinearProblemTest_MPI_4`

in the builds:

* `Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt`
* `Trilinos-atdm-waterman_cuda-9.2_shared_opt`
* `Trilinos-atdm-waterman-cuda-9.2-debug`
* `Trilinos-atdm-waterman-cuda-9.2-opt`
* `Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug`
* `Trilinos-atdm-waterman-cuda-9.2-release-debug`

started timing out every day on 2019-11-14.

The new commits that were pulled on 2019-11-14 shown [here](https://testing.sandia.gov/cdash/viewNotes.php?buildid=5902707#!#note3) involved a bunch of commits to upstream packages that could impact this test.  Note that as shown in [this query](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-10-22&end=2019-11-20&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman) there were no 'waterman' builds submitted on the days 2019-11-12 and 2019-11-13 because 'waterman' was being upgraded for the CUDA 10.1 driver.  So this set of commits is actually over 3 days of pushes to Trilinos 'develop'.

## Current Status on CDash

* ToDo: Fill in!

## Steps to Reproduce

One should be able to reproduce this failure on the machine <supported-atdm-machine> as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system <supported-atdm-system> are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#wwaterman

The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh \
    Trilinos-atdm-waterman-cuda-9.2-opt

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_ROL=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -n 20 ctest -j4
```

**Comment by bartlettroscoe** [comment]
Because of the upgrade of 'waterman' over those days, it is not clear if this test timeouts are due to a change in Trilinos or due to a change in the env with the upgrade of the CUDA drivers.  But given that this test is only timing out in the 'waterman' CUDA builds and not on builds on 'ride' or any other platforms, it seems likely this failure was triggered by the upgrade of the CUDA drivers.  But that does not mean that the underlying problem is not a defect in Trilinos.  It might just be that a latent defect in Trilinos is being expressed with the updated env.  It is hard to tell.  But there is lots of data on CDash that someone could use to help diagnose what might be happening.  Questions one can look on CDash for answers include:

1. Is this test slower in the other other CUDA builds as well on other platforms?  (This would show if the timeouts could be caused by a change in Trilinos or are likely due to the upgrade of the CUDA drivers on 'waterman'.)
1. Are there other tests on waterman that also got a lot slower between 2019-11-11 and 2019-11-14? (This could show if code built with CUDA 9.2 with the CUDA 10.1 drivers are slower than with the CUDA

**Comment by bartlettroscoe** [comment]
It occurred to me that we might try building from scratch on 'waterman' and see if this makes this test failure go away.  It may be that rebuilding may not be triggering the rebuilds of software that need to be rebuilt due to the upgrade of the nvcc compiler or the nvcc compiler may need to detect the new CUDA driver in order to build performant code?

**Comment by bartlettroscoe** [comment]
Looking at [all of the ROL_NonlinearProblemTest_MPI_4 runs over all of the ATDM Trilinos builds between 2019-10-22 and 2019-11-20](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-10-22&end=2019-11-20&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4) (which returned 332 test runs and took 2397.73s for CDash to bring up) and just in the the [cuda builds over that time period](https://testing.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-10-22&end=2019-11-20&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=buildname&compare2=63&value2=cuda&field3=testname&compare3=61&value3=ROL_NonlinearProblemTest_MPI_4), we can see that there is no sign that the runtimes for this test went up in any of the other builds, including the CUDA builds on 'ride'.

Therefore, I would suspect this is due to a system issue and not a code issue, but not sure to prove that is the case without some more work.

**Comment by bartlettroscoe** [comment]
Feedback from @fryeguy52 is that other teams are reporting slowdowns after the upgrade of waterman for the CUDA 10.1.z drivers.   Seems that some other updates were needed and seemed to fix problem (current on the 'dev' queue).  They should update the main compute nodes soon so we think we wait for this update and see if this fixes.

**Comment by bartlettroscoe** [comment]
FYI: As part of the work in to improve running of the Trilinos test suite on GPUs, it was found in https://github.com/trilinos/Trilinos/pull/6840#issuecomment-628732349 that this test does not pass when it is run by itself.  Therefore, adding the `RUN_SERIAL` will not make this test pass.

Therefore, this test need to just be disabled in all of the 'waterman' builds.

**Comment by bartlettroscoe** [comment]
As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-01&end=2020-05-21&filtercount=5&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=ATDM&field2=buildname&compare2=65&value2=Trilinos-atdm-&field3=buildname&compare3=64&value3=Trilinos-atdm-waterman&field4=testname&compare4=61&value4=ROL_NonlinearProblemTest_MPI_4&field5=testoutput&compare5=94&value5=Error%20initializing%20RM%20connection),, this test takes almost 6 minutes to complete in the 'vortex' CUDA builds.  It is just a very expensive tests in CUDA builds (which shows that perhaps this code is not really designed to run performantly in CUDA builds).

**Comment by bartlettroscoe** [comment]
This test is disabled in all ATDM Trilinos CUDA PR builds in PR #7407 and has been merged to 'atdm-nightly' in commit 4804b08.

Putting in review.

**Comment by bartlettroscoe** [comment]
CC: @rmmilewi 

Below is a sample of that comment content that will be added once a week to these issues in the first version of the tool being developed in #3887.  Several things need to be fixed before that can happen.

----

<html>

<body>

<h2>Test results for issue #6329 as of YYYY-MM-DD</h2>

<p>
<font color="gray">Tests with issue trackers Missing: twim=6</font><br>
</p>


<h3><font color="gray">Tests with issue trackers Missing: twim=6</font></h3>
<table style="width:100%" boarder="1">

<tr>
<th>Site</th>
<th>Build Name</th>
<th>Test Name</th>
<th>Status</th>
<th>Details</th>
<th>Consec&shy;utive Missing Days</th>
<th>Non-pass Last 30 Days</th>
<th>Pass Last 30 Days</th>
<th>Issue Tracker</th>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-debug&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman-cuda-9.2-debug</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">11</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-opt&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman-cuda-9.2-opt</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">11</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">12</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-release-debug&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman-cuda-9.2-release-debug</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">11</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-release-debug&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman_&shy;cuda-9.2_&shy;fpic_&shy;static_&shy;opt</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">12</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_fpic_static_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

<tr>
<td align="left">waterman</td>
<td align="left"><a href="https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=site&compare2=61&value2=waterman">Trilinos-atdm-waterman_&shy;cuda-9.2_&shy;shared_&shy;opt</a></td>
<td align="left">ROL_&shy;NonlinearProblemTest_&shy;MPI_&shy;4</td>
<td align="left"><font color="gray">Missing</font></td>
<td align="left">Missing</td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="gray">15</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="red">12</font></a></td>
<td align="right"><a href="https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2020-05-07&end=2020-06-05&filtercombine=and&filtercombine=&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman_cuda-9.2_shared_opt&field2=testname&compare2=61&value2=ROL_NonlinearProblemTest_MPI_4&field3=site&compare3=61&value3=waterman"><font color="green">0</font></a></td>
<td align="right"><a href="https://github.com/trilinos/Trilinos/issues/6329">#6329</a></td>
</tr>

</table>

</body>
</html>
